execution/gethexec/caff_node.go

```diff
@@ -0,0 +1,283 @@
+package gethexec
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"sync"
+	"time"
+
+	espressoClient "github.com/EspressoSystems/espresso-sequencer-go/client"
+
+	"github.com/ethereum/go-ethereum/arbitrum_types"
+	"github.com/ethereum/go-ethereum/core"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/rlp"
+
+	"github.com/offchainlabs/nitro/arbos"
+	"github.com/offchainlabs/nitro/arbos/arbostypes"
+	"github.com/offchainlabs/nitro/arbutil"
+	"github.com/offchainlabs/nitro/util/stopwaiter"
+)
+
+var (
+	retryTime = time.Second * 1
+)
+
+/*
+Caff Node creates blocks with finalized hotshot transactions
+*/
+type CaffNode struct {
+	stopwaiter.StopWaiter
+
+	config                  SequencerConfigFetcher
+	namespace               uint64
+	executionEngine         *ExecutionEngine
+	espressoClient          *espressoClient.Client
+	nextHotshotBlockNum     uint64
+	messagesWithMetadata    []*arbostypes.MessageWithMetadata
+	messagesWithMetadataPos []uint64
+	messagesStateMutex      sync.Mutex
+	skippedBlockPos         *uint64
+}
+
+func NewCaffNode(configFetcher SequencerConfigFetcher, execEngine *ExecutionEngine) *CaffNode {
+	config := configFetcher()
+	if err := config.Validate(); err != nil {
+		log.Crit("Failed to validate caff  node config", "err", err)
+	}
+	return &CaffNode{
+		config:              configFetcher,
+		namespace:           config.CaffNodeConfig.Namespace,
+		espressoClient:      espressoClient.NewClient(config.CaffNodeConfig.HotShotUrl),
+		nextHotshotBlockNum: config.CaffNodeConfig.StartBlock,
+		executionEngine:     execEngine,
+	}
+}
+
+// TODO: For future versions, we should check the attestation quote to check if its from a valid TEE
+// TODO: This machine should run in TEE and submit blocks to espresso only if the block is valid with an attestation.
+/**
+ * This function will create a block with the finalized hotshot transactions
+ * It will first remove duplicates and ensure the ordering of messages is correct
+ * Then it will run the STF using the `Produce Block`function and finally store the block in the database
+ */
+func (n *CaffNode) createBlock(ctx context.Context) (returnValue bool) {
+
+	n.messagesStateMutex.Lock()
+	defer n.messagesStateMutex.Unlock()
+
+	//  If we have no messages to process, return
+	if len(n.messagesWithMetadata) == 0 {
+		log.Warn("No messages to process")
+		return false
+	}
+	messageWithMetadata := n.messagesWithMetadata[0]
+	messageWithMetadataPos := n.messagesWithMetadataPos[0]
+
+	// Get the last block header stored in the database
+	if n.executionEngine.bc == nil {
+		log.Error("execution engine bc not initialized")
+		return false
+	}
+
+	lastBlockHeader := n.executionEngine.bc.CurrentBlock()
+
+	currentPos := lastBlockHeader.Number.Uint64()
```

Comment:
https://github.com/EspressoSystems/nitro-espresso-integration/blob/b9736855e6d72130a0175821703beff7cddeb2d0/execution/gethexec/executionengine.go#L644

I think we should use this function